src/Npgsql/NpgsqlTracingOptionsBuilder.cs

```diff
@@ -0,0 +1,112 @@
+    Func<NpgsqlBatch, string?>? _batchSpanNameProvider;
+    bool _enableFirstResponseEvent = true;
+
+    internal NpgsqlTracingOptionsBuilder()
+    {
+    }
+
+    /// <summary>
+    /// Configures a filter function that determines whether to emit tracing information for an <see cref="NpgsqlCommand"/>.
+    /// By default, tracing information is emitted for all commands.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder ConfigureCommandFilter(Func<NpgsqlCommand, bool>? commandFilter)
+    {
+        _commandFilter = commandFilter;
+        return this;
+    }
+
+    /// <summary>
+    /// Configures a filter function that determines whether to emit tracing information for an <see cref="NpgsqlBatch"/>.
+    /// By default, tracing information is emitted for all batches.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder ConfigureBatchFilter(Func<NpgsqlBatch, bool>? batchFilter)
+    {
+        _batchFilter = batchFilter;
+        return this;
+    }
+
+    /// <summary>
+    /// Configures a callback that can enrich the <see cref="Activity"/> emitted for the given <see cref="NpgsqlCommand"/>.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder ConfigureCommandEnrichmentCallback(Action<Activity, NpgsqlCommand>? commandEnrichmentCallback)
+    {
+        _commandEnrichmentCallback = commandEnrichmentCallback;
+        return this;
+    }
+
+    /// <summary>
+    /// Configures a callback that can enrich the <see cref="Activity"/> emitted for the given <see cref="NpgsqlBatch"/>.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder ConfigureBatchEnrichmentCallback(Action<Activity, NpgsqlBatch>? batchEnrichmentCallback)
+    {
+        _batchEnrichmentCallback = batchEnrichmentCallback;
+        return this;
+    }
+
+    /// <summary>
+    /// Configures a callback that provides the tracing span's name for an <see cref="NpgsqlCommand"/>. If <c>null</c>, the default standard
+    /// span name is used, which is the database name.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder ConfigureCommandSpanNameProvider(Func<NpgsqlCommand, string?>? commandSpanNameProvider)
+    {
+        _commandSpanNameProvider = commandSpanNameProvider;
+        return this;
+    }
+
+    /// <summary>
+    /// Configures a callback that provides the tracing span's name for an <see cref="NpgsqlBatch"/>. If <c>null</c>, the default standard
+    /// span name is used, which is the database name.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder ConfigureBatchSpanNameProvider(Func<NpgsqlBatch, string?>? batchSpanNameProvider)
+    {
+        _batchSpanNameProvider = batchSpanNameProvider;
+        return this;
+    }
+
+    /// <summary>
+    /// Gets or sets a value indicating whether to enable the "time-to-first-read" event.
+    /// Default is true to preserve existing behavior.
+    /// </summary>
+    public NpgsqlTracingOptionsBuilder EnableFirstResponseEvent(bool enable = true)
```

Comment:
For things that are enabled by default, has this pattern been confusing to EF users? Where they have to use the enable method to disable a default?

Reply:
I understand the hesitation - I haven't seen anybody actually have trouble with it... People who don't want to disable never see it or care, and people who do figure it out...

I'd be OK with flipping this around (DisableFirstResponseEvent), but then you get double-negation passing false, so... Let me know what you prefer.

Reply:
Yeah I would definitely stick to the non negated naming convention. Was wondering whether you had gotten any EF feedback calling out (somewhat) worse discoverability.

Reply:
Nope, not really...